consensus/hotstuff/signature.go

```diff
@@ -107,6 +107,43 @@ type WeightedSignatureAggregator interface {
 	Aggregate() ([]flow.Identifier, []byte, error)
 }
 
+// MultiMessageSignatureAggregator aggregates signatures of the same signature scheme but
+// possibly distinct message from different signers. Only public keys are agreed upon upfront.
+// It's similar to WeightedSignatureAggregator but supports distinct messages.
+// It is also recommended to only aggregate signatures generated with keys representing
+// equivalent security-bit level.
+// Furthermore, a weight [unsigned int64] is assigned to each signer ID. The
+// MultiMessageSignatureAggregator internally tracks the total weight of all collected signatures.
+// Implementations must be concurrency safe.
+type MultiMessageSignatureAggregator interface {
+	// Verify verifies the signature under the stored public keys.
+	// Expected errors during normal operations:
+	//  - model.InvalidSignerError if signerID is invalid (not a consensus participant)
+	//  - model.ErrInvalidSignature if signerID is valid but signature is cryptographically invalid
+	Verify(signerID flow.Identifier, sig crypto.Signature, msg []byte) error
+
+	// TrustedAdd adds signature and message to the internal set of signatures and adds the signer's
+	// weight to the total collected weight, iff the signature is _not_ a duplicate.
+	// The total weight of all collected signatures (excluding duplicates) is returned regardless
+	// of any returned error.
+	// The function errors with:
+	//  - model.InvalidSignerError if signerID is invalid (not a consensus participant)
+	//  - model.DuplicatedSignerError if the signer has been already added
+	TrustedAdd(signerID flow.Identifier, sig crypto.Signature, msg []byte) (totalWeight uint64, exception error)
+
+	// TotalWeight returns the total weight presented by the collected signatures.
+	TotalWeight() uint64
+
+	// Aggregate aggregates the signatures and returns the aggregated signature.
+	// The function performs a final verification and errors if the aggregated
+	// signature is not valid. This is required for the function safety since
+	// `TrustedAdd` allows adding invalid signatures.
+	// Expected errors during normal operations:
+	//  - model.InsufficientSignaturesError if no signatures have been added yet
+	//  - model.InvalidSignatureIncludedError if some signature(s), included via TrustedAdd, are invalid
+	Aggregate() ([]flow.Identifier, []byte, error)
```

Comment:
I didn't include messages, since other component will be building data for reconstructing messages. In other words I don't really need it. If you think we need to add it for general usage we can add it.

Reply:
We could:
 - either return `signers map[flow.Identifier][]byte , sig Signature, error`  for generic use cases. 
 - OR only return `sig Signature, error` because the list of signers in an array doesn't add any information about the aggregated signature.
 
 > since other component will be building data for reconstructing messages.
 So when the aggregator node broadcasts/sends this aggregated signature to other nodes, it doesn't need to also mention the list of signers and their messages? 
(FYI, I know nothing about the pace maker design)

Reply:
We have a mixed case, providing some context, we have a timeout certificate, which is similar to quorum certificate but has multiple messages. 

```golang
type TimeoutCertificate struct {
	View      uint64
	SignerIDs []Identifier
	TOHighQCViews []uint64 // highest QC view
	TOHighestQC *flow.QuorumCertificate // highest QC built based on TOHighQCViews
	SigData   []byte // Aggregated signature over multiple messages, 
									 // one message consist of (View, HighestQCView)
}
```

This structure will be filled after aggregation process. 
`SigData` - is the aggregated signature itself
`SignerIDs` - is the list of signers contributed to creating TC
`TOHighQCViews` - is the array of views which is used to create the messages for sig verification(also used for other validity checks)

So when receiving TC to verify `SigData` we will reconstruct messages using `View` and `TOHighQCViews`

Reply:
Thanks for the context, I understand better now. 

The module's API itself looks fine, but if we look at it as part of the wider design, there might be a conceptual safety issue. This module collects signatures and messages via `TrustedAdd`, I assume each signature has been already verified using `Verify` (like in the QC case). In order to guarantee that the aggregated signature at the end is safe, we execute an internal extra verification of the aggregated signature. That verification is there as a safety measure to make sure calls to `TrustedAdd` and `Verify` are consistent. The last verification guarantees that the aggregated signature is valid against the **set of messages collected internally** (and not another set of messages). However, the design seems to be collection the messages in another module (the one collecting  `View` and `TOHighQCViews`). 
The extra verification inside `SignatureAggregator` doesn't add much to the validity of the aggregated signature against `TOHighQCViews`.

Still thinking about the best way to have a safe design 🤔

Reply:
Note that the the last sanity verification (different messages) isn't cheap like the one in QC. Some options to address the issue above:

1. run another safety verification of the aggregated signature against a message derived from `View` and `TOHighQCViews`, when creating `TimeoutCertificate`.  [expensive option because the safety verification is run twice]
2. rename `Aggregate` to `UnsafeAggregate` and remove the first safety verification from the module. Keep the second verification when creating `TimeoutCertificate`. 
3. Update the the aggregator to accept `TOHighQCView uint64` instead of `message []byte` in `Verify` and `TrustedAdd`. `Aggregate` would return `signature crypto.Signature, SignerIDs []Identifier, TOHighQCViews []uint64` that are all safe to go into `TimeoutCertificate`.

Reply:
What if we don't cache messages and pass them in `Aggregate` as parameter?
On other side I am tempted to do `UnsafeAggregate` and then validate resulting TC(not only signature but whole TC).

Reply:
> What if we don't cache messages and pass them in Aggregate as parameter

that's also possible

Reply:
I think it might be helpful here to consider the reasons why we have the `TrustedAdd` in the `WeightedSignatureAggregator` (same message):
* **Vote Aggregation**: same message, but potentially two different signatures from the same voter:
   
   This is the scenario  for which we designed `WeightedSignatureAggregator`. Votes (depending on the voting scheme) could contain a staking _and_ a beacon signatures, which must either _both_ be added or not added. Therefore, we have _two_ aggregators, whose internal state needs to be evolved in synchrony, ideally without too much locking.  
   * Separating `Verify` and `TrustedAdd` allowed us to _safely_ address this case in a performant manner, because we can verify both signatures upfront. In other words, we check that it is safe to add the votes upfront, before we commence any state-changing operations. 
   * As @tarakby explained, the `WeightedSignatureAggregator` does a posterior validation of the signature aggregate _before_ it returns the result. This protects from accidental misuses, where a faulty signature is added via `TrustedAdd`. So `WeightedSignatureAggregator` guarantees correctness for any aggregate it returns (a very important property!). 
     When aggregating signatures over the _same_ message, the posterior validation adds only marginal overhead (essentially 1 extra validation on top of the 90+ validations we anyway do using `Verify`). 
* **Timeout Aggregation**: each signer potentially signs a different message, but each signer only provides a _single_ signature:

  This is the scenario  for which we designing `WeightedMultiMessageSignatureAggregator`. While it would be nice to have consistent API, a safety guarantee is much more important.  

  * Note that in this case, there is only a single struct that carries the relevant state: a single instance of `WeightedMultiMessageSignatureAggregator`. Therefore, we could work with the method
    ```golang
    VerifyAndAdd(signerID flow.Identifier, sig crypto.Signature, msg []byte)
    ```
  * We can design the code such that congestion is not a problem, if we check signature validity internally _before_ we lock the internal state. 
  * With `VerifyAndAdd`, the aggregator can itself assert that all input signatures are valid. Therefore, the aggregator  can guarantee correctness for any aggregate it returns without posterior validation of the result. 
    For this scenario, posterior validation is much more expensive because different messages are signed. 

In summary, I would suggest to favour safety over API consistency

Reply:
Thanks @AlexHentschel for the context.

It looks like the use cases of `WeightedMultiMessageSignatureAggregator` and `WeightedSignatureAggregator` are different. After the context you added, I agree there is no need for `WeightedMultiMessageSignatureAggregator` to separate `Add` and `Verify` and we could have these in a single step `VerifyAndAdd`.  In that case `Aggregate` is a safe function without the need to add a final verification. 

The wider safety issue still remains though, and there is final step of verification needed: the signature returned by `WeightedMultiMessageSignatureAggregator` is safe against the messages collected internally by the module, but is not safe against the `TOHighQCViews`, which are collected by a separate module. 

Some options to fix this are:

1. run another safety verification of the aggregated signature against a message derived from View and `TOHighQCViews`, when creating `TimeoutCertificate`. [run an expensive safety verification].
2. Update the the aggregator to accept `TOHighQCView uint64` instead of `message []byte` in `VerifyAndAdd`. `Aggregate` would return `signature crypto.Signature, SignerIDs []Identifier, TOHighQCViews []uint64` that are all safe to go into TimeoutCertificate.

Reply:
thanks Tarak for the insights. 

> 2. Update the the aggregator to accept TOHighQCView uint64 instead of message []byte in VerifyAndAdd. Aggregate would return signature crypto.Signature, SignerIDs []Identifier, TOHighQCViews []uint64 that are all safe to go into TimeoutCertificate.

this would be my preference